core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -223,9 +227,24 @@ public ShardFailedClusterStateTaskExecutor(AllocationService allocationService,
         @Override
         public BatchResult<ShardRoutingEntry> execute(ClusterState currentState, List<ShardRoutingEntry> tasks) throws Exception {
             BatchResult.Builder<ShardRoutingEntry> batchResultBuilder = BatchResult.builder();
+            Set<ShardRoutingEntry> nonTrivialTasks = Collections.newSetFromMap(new IdentityHashMap<>());
             List<FailedRerouteAllocation.FailedShard> failedShards = new ArrayList<>(tasks.size());
             for (ShardRoutingEntry task : tasks) {
-                failedShards.add(new FailedRerouteAllocation.FailedShard(task.shardRouting, task.message, task.failure));
+                RoutingNodes.RoutingNodeIterator routingNodeIterator =
+                    currentState.getRoutingNodes().routingNodeIter(task.getShardRouting().currentNodeId());
+                if (routingNodeIterator != null) {
+                    for (ShardRouting maybe : routingNodeIterator) {
+                        if (task.getShardRouting().isSameAllocation(maybe)) {
+                            nonTrivialTasks.add(task);
+                            failedShards.add(new FailedRerouteAllocation.FailedShard(task.shardRouting, task.message, task.failure));
+                            break;
+                        }
+                    }
+                }
+                if (!nonTrivialTasks.contains(task)) {
```

Comment:
I simplified and clarified the logic in fc014e691521e1e15d2ec3f0804aae619318e461.